modules/client-handler/src/main/java/org/apache/ignite/client/handler/JdbcQueryEventHandlerImpl.java

```diff
@@ -369,56 +380,75 @@ private CompletionStage<JdbcQuerySingleResult> createJdbcResult(AsyncSqlCursor<I
         return cur.requestNextAsync(req.pageSize()).thenApply(batch -> {
             boolean hasNext = batch.hasMore();
 
+            long cursorId;
+            try {
+                cursorId = resources.put(new ClientResource(cur, cur::closeAsync));
+            } catch (IgniteInternalCheckedException e) {
+                cur.closeAsync();
+
+                return new JdbcQuerySingleResult(Response.STATUS_FAILED,
+                        "Unable to store query cursor.");
+            }
+
             switch (cur.queryType()) {
                 case EXPLAIN:
                 case QUERY: {
-                    long cursorId;
-                    try {
-                        cursorId = resources.put(new ClientResource(cur, cur::closeAsync));
-                    } catch (IgniteInternalCheckedException e) {
-                        cur.closeAsync();
-
-                        return new JdbcQuerySingleResult(Response.STATUS_FAILED,
-                                "Unable to store query cursor.");
-                    }
-
-                    List<BinaryTupleReader> rows = new ArrayList<>(batch.items().size());
-                    for (InternalSqlRow item : batch.items()) {
-                        rows.add(item.asBinaryTuple());
-                    }
-
                     List<ColumnMetadata> columns = cur.metadata().columns();
 
-                    int[] decimalScales = new int[columns.size()];
-                    List<ColumnType> schema = new ArrayList<>(columns.size());
-
-                    int countOfDecimal = 0;
-                    for (ColumnMetadata column : columns) {
-                        schema.add(column.type());
-                        if (column.type() == ColumnType.DECIMAL) {
-                            decimalScales[countOfDecimal++] = column.scale();
-                        }
-                    }
-                    decimalScales = Arrays.copyOf(decimalScales, countOfDecimal);
-
-                    return new JdbcQuerySingleResult(cursorId, rows, schema, decimalScales, !hasNext);
+                    return buildSingleRequest(batch, columns, cursorId, !hasNext, cur.queryType());
                 }
-                case DML:
+                case DML: {
```

Comment:
if DML is handled by `buildSingleRequest` as well, then we should merge these branches